File: apps/frontend/src/api/documents.ts
import type { DuplicatePoint, NormalizedDocument } from "@jhl548/duplicate-doc-vue";
import type { EditorChangePayload } from "@jhl548/duplicate-doc-vue";

export interface AnalyzeResponse {
  mainDocumentId: string;
  documents: NormalizedDocument[];
  duplicates: DuplicatePoint[];
}

export interface ApiResponse<T> {
  code: number;
  message: string;
  data: T;
  meta?: {
    traceId?: string;
    timestamp?: string;
  };
}

export interface ProjectDemoViewData extends AnalyzeResponse {
  project: {
    projectId: string;
    name: string;
    mainDocumentId: string;
  };
  activeDefaults: {
    duplicateId?: string | null;
    slaveDocumentId?: string | null;
  };
  stats: {
    documentTotal: number;
    duplicateTotal: number;
    rangeMapTotal: number;
    plainTextTotal: number;
  };
  capabilities: {
    supportedFileTypes: string[];
    supportedContentTypes: string[];
  };
  actions: Record<
    string,
    {
      method: string;
      url: string;
      description: string;
    }
  >;
}

export interface EditSnapshotRequest extends EditorChangePayload {
  fileName: string;
}

export interface EditSnapshotResponse {
  documentId: string;
  fileName: string;
  htmlLength: number;
  plainTextLength: number;
  accepted: boolean;
}

export async function analyzeDocuments(mainFile: File, slaveFiles: File[]): Promise<ProjectDemoViewData> {
  const formData = new FormData();
  formData.append("main_file", mainFile);
  for (const file of slaveFiles) {
    formData.append("slave_files", file);
  }

  const response = await fetch("/api/projects/analyze-view", {
    method: "POST",
    body: formData
  });

  if (!response.ok) {
    throw new Error(await response.text());
  }

  const result = (await response.json()) as ApiResponse<ProjectDemoViewData>;
  if (result.code !== 0) {
    throw new Error(result.message || "分析失败");
  }

  return result.data;
}

export async function exportDocx(documentId: string, fileName: string, html: string): Promise<void> {
  const response = await fetch("/api/documents/export-docx", {
    method: "POST",
    headers: {
      "Content-Type": "application/json"
    },
    body: JSON.stringify({ documentId, fileName, html })
  });

  if (!response.ok) {
    throw new Error(await response.text());
  }

  const blob = await response.blob();
  const url = URL.createObjectURL(blob);
  const link = window.document.createElement("a");
  link.href = url;
  link.download = fileName.endsWith(".docx") ? fileName : `${fileName}.docx`;
  link.click();
  URL.revokeObjectURL(url);
}

export async function submitEditedSnapshot(payload: EditSnapshotRequest): Promise<EditSnapshotResponse> {
  const response = await fetch("/api/documents/edit-snapshot", {
    method: "POST",
    headers: {
      "Content-Type": "application/json"
    },
    body: JSON.stringify(payload)
  });

  if (!response.ok) {
    throw new Error(await response.text());
  }

  return response.json() as Promise<EditSnapshotResponse>;
}

File: apps/backend/app/schemas/documents.py
from typing import Any, Literal

from pydantic import BaseModel, Field


DocumentRegion = Literal["body", "header", "footer", "table", "caption", "unknown"]
DuplicateSeverity = Literal["noise", "low", "medium", "high"]
TenderSemanticType = Literal[
    "qualification",
    "priceTable",
    "deviationTable",
    "personnel",
    "performance",
    "schedule",
    "technical",
    "business",
    "projectInfo",
    "unknown",
]


class TableCellContext(BaseModel):
    tableId: str
    rowIndex: int = Field(ge=0)
    cellIndex: int = Field(ge=0)
    headerText: str | None = None
    rowHeaderText: str | None = None


class TextRange(BaseModel):
    start: int = Field(ge=0)
    end: int = Field(ge=0)
    blockId: str | None = None
    sectionPath: list[str] | None = None
    region: DocumentRegion | None = None
    tableContext: TableCellContext | None = None
    confidence: float | None = Field(default=None, ge=0, le=1)
    semanticType: TenderSemanticType | None = None


class RangeMapEntry(BaseModel):
    blockId: str
    textStart: int = Field(ge=0)
    textEnd: int = Field(ge=0)
    selector: str | None = None
    sectionPath: list[str] | None = None
    region: DocumentRegion | None = None
    listMarker: str | None = None
    tableContext: TableCellContext | None = None
    semanticType: TenderSemanticType | None = None


class DocumentAsset(BaseModel):
    id: str
    url: str
    type: Literal["image", "font", "other"] = "other"


class NormalizedDocument(BaseModel):
    documentId: str
    role: Literal["main", "slave"]
    fileName: str
    html: str
    plainText: str
    rangeMap: list[RangeMapEntry]
    assets: list[DocumentAsset] = Field(default_factory=list)
    meta: dict[str, Any] = Field(default_factory=dict)


class DuplicateDocumentRanges(BaseModel):
    documentId: str
    ranges: list[TextRange]


class DuplicatePoint(BaseModel):
    duplicateId: str
    groupId: str | None = None
    similarity: float = Field(ge=0, le=1)
    label: str | None = None
    summary: str | None = None
    reason: str | None = None
    ignored: bool | None = None
    severity: DuplicateSeverity | None = None
    semanticType: TenderSemanticType | None = None
    region: DocumentRegion | None = None
    noiseReason: str | None = None
    main: DuplicateDocumentRanges
    slaves: list[DuplicateDocumentRanges]


class AnalyzeResponse(BaseModel):
    mainDocumentId: str
    documents: list[NormalizedDocument]
    duplicates: list[DuplicatePoint]


class ProjectInfo(BaseModel):
    projectId: str
    name: str
    mainDocumentId: str


class ActiveDefaults(BaseModel):
    duplicateId: str | None = None
    slaveDocumentId: str | None = None


class ProjectStats(BaseModel):
    documentTotal: int
    duplicateTotal: int
    rangeMapTotal: int
    plainTextTotal: int


class ProjectCapabilities(BaseModel):
    supportedFileTypes: list[str]
    supportedContentTypes: list[str]


class ProjectAction(BaseModel):
    method: str
    url: str
    description: str


class ProjectDemoViewData(BaseModel):
    project: ProjectInfo
    mainDocumentId: str
    documents: list[NormalizedDocument]
    duplicates: list[DuplicatePoint]
    activeDefaults: ActiveDefaults
    stats: ProjectStats
    capabilities: ProjectCapabilities
    actions: dict[str, ProjectAction]


class ApiMeta(BaseModel):
    traceId: str | None = None
    timestamp: str | None = None


class ProjectDemoViewResponse(BaseModel):
    code: int = 0
    message: str = "ok"
    data: ProjectDemoViewData
    meta: ApiMeta = Field(default_factory=ApiMeta)


class ExportRequest(BaseModel):
    documentId: str
    fileName: str = "edited-document.docx"
    html: str


class EditSnapshotRequest(BaseModel):
    documentId: str
    fileName: str = "edited-document.docx"
    html: str
    plainText: str
    json: Any | None = None


class EditSnapshotResponse(BaseModel):
    documentId: str
    fileName: str
    htmlLength: int
    plainTextLength: int
    accepted: bool = True
